crates/supervisor/storage/src/providers/derivation_provider.rs

```diff
@@ -263,18 +304,69 @@ where
                 );
             })?;
 
-        // Save the derived block numbers to the database
-        self.tx
-            .put::<SourceToDerivedBlockNumbers>(incoming_pair.source.number, derived_block_numbers)
-            .inspect_err(|err| {
+        // Save the SourceBlockTraversal to the database
+        self.tx.put::<BlockTraversal>(incoming_pair.source.number, block).inspect_err(|err| {
+            error!(
+                target: "supervisor_storage",
+                incoming_derived_block_pair = %incoming_pair,
+                %err,
+                "Failed to save derived block numbers for source block"
+            );
+        })?;
+
+        Ok(())
+    }
+
+    /// Saves a source block to the database.
+    /// If the source block already exists, it does nothing.
+    /// If the source block does not exist, it creates a new [`SourceBlockTraversal`] and saves it
+    /// to the database.
+    pub(crate) fn save_source_block(
+        &self,
+        source: BlockInfo,
+    ) -> Result<SourceBlockTraversal, StorageError> {
+        match self.get_block_traversal(source.number) {
```

Comment:
We don't need to perform 2 fetch here `get_block_traversal` and `latest_source_block`. 

We can just fetch `latest_source_block` and do the following
- check if it is same return Ok
- if incoming block is child of `latest_source_block`, then save the incoming
- else error